# Исследование надежности заемщиков

## Изучение общей информации о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

Последовательно импортируем нужную библиотеку и загрузим исходный файл:

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

Выведем нужные строки. Отметим, что столбец days_employed имеет явно аномальные (отрицательные и очень большие значения):

In [3]:
data.head(20)

Unnamed: 0  children  days_employed  dob_years            education  \
0            0         1   -8437.673028         42               высшее   
1            1         1   -4024.803754         36              среднее   
2            2         0   -5623.422610         33              Среднее   
3            3         3   -4124.747207         32              среднее   
4            4         0  340266.072047         53              среднее   
5            5         0    -926.185831         27               высшее   
6            6         0   -2879.202052         43               высшее   
7            7         0    -152.779569         50              СРЕДНЕЕ   
8            8         2   -6929.865299         35               ВЫСШЕЕ   
9            9         0   -2188.756445         41              среднее   
10          10         2   -4171.483647         36               высшее   
11          11         0    -792.701887         40              среднее   
12          12         0            NaN         65              среднее   
13          13         0   -1846.641941         54  неоконченное высшее   
14          14         0   -1844.956182         56               высшее   
15          15         1    -972.364419         26              среднее   
16          16         0   -1719.934226         35              среднее   
17          17         0   -2369.999720         33               высшее   
18          18         0  400281.136913         53              среднее   
19          19         0  -10038.818549         48              СРЕДНЕЕ   

    education_id     family_status  family_status_id gender income_type  debt  \
0              0   женат / замужем                 0      F   сотрудник     0   
1              1   женат / замужем                 0      F   сотрудник     0   
2              1   женат / замужем                 0      M   сотрудник     0   
3              1   женат / замужем                 0      M   сотрудник     0   
4              1  гражданский брак                 1      F   пенсионер     0   
5              0  гражданский брак                 1      M   компаньон     0   
6              0   женат / замужем                 0      F   компаньон     0   
7              1   женат / замужем                 0      M   сотрудник     0   
8              0  гражданский брак                 1      F   сотрудник     0   
9              1   женат / замужем                 0      M   сотрудник     0   
10             0   женат / замужем                 0      M   компаньон     0   
11             1   женат / замужем                 0      F   сотрудник     0   
12             1  гражданский брак                 1      M   пенсионер     0   
13             2   женат / замужем                 0      F   сотрудник     0   
14             0  гражданский брак                 1      F   компаньон     1   
15             1   женат / замужем                 0      F   сотрудник     0   
16             1   женат / замужем                 0      F   сотрудник     0   
17             0  гражданский брак                 1      M   сотрудник     0   
18             1    вдовец / вдова                 2      F   пенсионер     0   
19             1         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   7

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

Из основной информации можем выявить общий размер датов и наличие пропусков в столбцах days_employed и total_income:

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        21525 non-null  int64  
 1   children          21525 non-null  int64  
 2   days_employed     19351 non-null  float64
 3   dob_years         21525 non-null  int64  
 4   education         21525 non-null  object 
 5   education_id      21525 non-null  int64  
 6   family_status     21525 non-null  object 
 7   family_status_id  21525 non-null  int64  
 8   gender            21525 non-null  object 
 9   income_type       21525 non-null  object 
 10  debt              21525 non-null  int64  
 11  total_income      19351 non-null  float64
 12  purpose           21525 non-null  object 
dtypes: float64(2), int64(6), object(5)
memory usage: 2.1+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

Посчитаем количество пропущенных значений для каждого столбца. В столбцах days_employed и total_income имеется по 2174 пропуска:

In [5]:
data.isna().sum()

Unnamed: 0             0
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

Заменим пропуски предложенным способом:

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

Заменим аномалии требуемым способом:

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

С помощью метода groupby выведем требуемые значения:

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

Среди уникальных значений обратим внимание на аномальные значения -1 и 20:

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

Действительно, мы верно удалили аномальные значения:

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

Пропущенных значений больше нет, мы всё сделали корректно:

In [13]:
data.isna().sum()

Unnamed: 0          0
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Заменить вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработать неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Следует привести их к нижнему регистру и проверить остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Вывести на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалить их.**

Посчитаем дубликаты и убедимся, что дубликаты отсутствуют:

In [16]:
data.duplicated().sum()

0

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создать в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Следут использовать собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Вывести на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создать функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Следует использовать собственную функцию с именем `categorize_purpose()` и метод `apply()`.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных и ответ на вопросы

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

Прежде всего, посчитаем долю просрочивших для каждого варианта количества детей:

In [22]:
data.groupby('children')['debt'].mean()

children
0    0.075129
1    0.092154
2    0.094404
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

Отсутствие просрочивших среди дебиторов с пятью детьми, вероятно, объясняется тем, что их просто слишком мало в наборе данных, чтобы обнаружить хотя бы одного просрочившего. Проверим это предположение, добавиви в таблицу их точное количество:

In [23]:
data.groupby('children')['debt'].agg(['count', 'mean'])

count      mean
children                 
0         14149  0.075129
1          4818  0.092154
2          2055  0.094404
3           330  0.081818
4            41  0.097561
5             9  0.000000

Действительно, дебиторов с пятью детьми лишь девять человек, а в среднем на одного просрочившего приходится более десяти выплативших кредит вовремя.

**Вывод:** Судя по всему, на возврат кредита в срок влияет не количество детей, а их наличие. Так, среди тех, у кого нет детей, процент должников составлял порядка 7.5%, а среди остальных - от 8.2% до 9.8%.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Прежде всего, посчитаем долю просрочивших для каждого семейного статуса (чтобы выявить слишком малые категории, вновь добавим в таблицу общее количество дебиторов каждой категории):

In [24]:
data.groupby('family_status')['debt'].agg(['count', 'mean']).sort_values('mean', ascending=False)

count      mean
family_status                         
Не женат / не замужем   2799  0.097535
гражданский брак        4160  0.092548
женат / замужем        12302  0.075354
в разводе               1189  0.070648
вдовец / вдова           952  0.066176

**Вывод:** Процент просрочивших выше среди тех, кто никогда не состоял в официальном браке (неженатые и состоящие в гражданском браке) и составляет 9.3%-9.8%. Среди тех, кто когда-либо состоял в браке (женатые, вдовцы и разведённые) процент должников ниже и составляет 6.6%-7.6%.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Прежде всего, посчитаем долю просрочивших для каждой категории уровня доходов и общее количество дебиторов в ней:

In [25]:
data.groupby('total_income_category')['debt'].agg(['count', 'mean'])

count      mean
total_income_category                 
A                         25  0.080000
B                       5014  0.070602
C                      15992  0.084605
D                        349  0.060172
E                         22  0.090909

**Вывод:** Среди результатов выделяются самая низкая доля просрочивших среди людей с категорией дохода D (6%) и самая высокая в категории E (9.1%). Однако в целом предположение, что чем больше доход, тем меньше доля просрочивших, не подтверждается.

### Как разные цели кредита влияют на его возврат в срок?

Прежде всего, посчитаем долю просрочивших для каждой категории целей кредита и общее количество дебиторов в ней:

In [26]:
data.groupby('purpose_category')['debt'].agg(['count', 'mean']).sort_values('mean', ascending=False)

count      mean
purpose_category                         
операции с автомобилем     4288  0.093284
получение образования      3997  0.092319
проведение свадьбы         2337  0.078306
операции с недвижимостью  10780  0.072356

**Вывод:** Меньше всего просрочивших среди тех, кто брал кредит на операции с недвижимостью (7.3%) и проведение свадьбы (7.9%). Несколько выше эта доля среди взявших в кредит на операции с автомобилем или получение образования (9.3%).

### Каковы возможные причины появления пропусков в исходных данных?

*Ответ:* Судя по тому, что пропуски возникают исключительно в двух конкретных столбцах и всегда одновременно, эти пропуски не являются полностью случайными. Нет и причин считать, что вероятность пропусков зависит от данных в других столбцах, то есть пропуски вряд ли случайные. Могу предположить, пропуски неслучайны и некоторые люди намеренно не указали свой заработок и трудовой стаж, сочтя это слишком конфиденциальной информацией или чтобы не снижать шанс на кредит.

### Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?

*Ответ:* При заполнении пропусков следует использовать наиболее характерные значения относительно имеющихся данных; обычно это среднее или медиана. В данном случае количественные переменные с пропусками - это ежемесячный доход и общий трудовой стаж, то есть величины с большим разбросом в значениях. В этой ситуации взятие среднего приведёт к сдвигу, так как, например, малое количество людей, чей заработок очень велик, сдвинут среднее значение в большую сторону, хотя это никак не будет характеризовать основную часть данных. Для таких переменных, как возраст, где обычно нет такого разброса данных, вполне годится и среднее.

## Общий вывод

> В рамках проекта мной исследовалась статистика платёжеспособности клиентов на основе набора данных из 21525 вхождений, описывающих дебиторов. Работа велась с такими признаками, как количество детей, трудовой стаж, семейное положение, тип занятости, ежемесячный доход, цель получения кредита. Целевым признаком, относительно которого и выявлялись зависимости, был факт наличия задолженности по ходу выплаты кредита.  
  
> В ходе предобработки были обнаружены аномальные отрицательные значения трудового стажа, которые было решено заменить на аналогичные, но положительные. Также из набора были удалены дебиторы с аномальным количеством детей -1 и 20. Пропуски были обнаружены только в столбцах с доходом и стажем, они были заполнены медианными значениями для соответствующего типа занятости. Затем были удалены все дубликаты в количестве 71, а в столбце дохода оставлена лишь целая часть.  В завершение, была проведена категоризация данных по доходу (на 5 категорий с порогами в 30 тыс., 50 тыс., 200 тыс. и 1000 тыс.) и по целям кредита (на 4 категории - операции с недвижимостью, с автомобилями, получение образование, проведение свадьбы).  
  
> Наконец, по имеющимся данным удалось выделить следующие факторы, повышающие риск того, что кредит не будет возвращён в срок: наличие детей, отсутствие опыта брака, доход менее 30000 рублей и направленность кредита на покупку машины или получения образования. Вероятно, именно эти показатели должны учитываться системой скоринга для выявления потенциальных заёмщиков с низкой способностью возвращения кредита.  
  
> В дальнейшем при сборе данных можно порекомендовать тщательнее подходить к заполнению стажа работы и заработка, чтобы не допускать отрицательных и минимизировать пропущенные значения. Также, в будущем может быть полезно исследовать влияние таких параметров, как образование и пол дебитора, на возврат кредита в срок для улучшения системы скоринга.